File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/D/ReleaseLock.cs
// -----------------------------------------------------------------------
// <copyright file="ReleaseLock.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.D;

using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using H = D2.Shared.Interfaces.Caching.Distributed.Handlers.D.IDelete.IReleaseLockHandler;
using I = D2.Shared.Interfaces.Caching.Distributed.Handlers.D.IDelete.ReleaseLockInput;
using O = D2.Shared.Interfaces.Caching.Distributed.Handlers.D.IDelete.ReleaseLockOutput;

/// <summary>
/// Handler for releasing a distributed lock in Redis using an atomic Lua compare-and-delete.
/// </summary>
public partial class ReleaseLock : BaseHandler<H, I, O>, H
{
    /// <summary>
    /// Lua script for atomic compare-and-delete.
    /// Only deletes the key if the stored value matches the provided lock ID.
    /// </summary>
    private const string _RELEASE_LOCK_SCRIPT = """
        if redis.call("get", KEYS[1]) == ARGV[1] then
            return redis.call("del", KEYS[1])
        else
            return 0
        end
        """;

    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="ReleaseLock"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public ReleaseLock(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        try
        {
            var db = r_redis.GetDatabase();
            var result = await db.ScriptEvaluateAsync(
                _RELEASE_LOCK_SCRIPT,
                [(RedisKey)input.Key],
                [(RedisValue)input.LockId]);

            var released = (int)result == 1;
            return D2Result<O?>.Ok(new O(released));
        }
        catch (RedisException ex)
        {
            LogReleaseLockFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<O?>.ServiceUnavailable(traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while releasing a distributed lock.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while releasing lock for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogReleaseLockFailed(ILogger logger, Exception ex, string key, string? traceId);
}

File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/d/release-lock.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";

type Input = DistributedCache.ReleaseLockInput;
type Output = DistributedCache.ReleaseLockOutput;

/** Lua script for atomic compare-and-delete. Only deletes if the stored value matches lockId. */
const _RELEASE_LOCK_SCRIPT = `
if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("del", KEYS[1])
else
    return 0
end
`;

export class ReleaseLock
  extends BaseHandler<Input, Output>
  implements DistributedCache.IReleaseLockHandler
{
  private readonly redis: Redis;

  constructor(redis: Redis, context: IHandlerContext) {
    super(context);
    this.redis = redis;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    try {
      const result = await this.redis.eval(_RELEASE_LOCK_SCRIPT, 1, input.key, input.lockId);
      return D2Result.ok({ data: { released: result === 1 } });
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}
